src/agent.rs

```diff
@@ -187,4 +210,130 @@ impl EmulatorCoreAgentState {
             (1000 / self.speed).into()
         }
     }
+
+    pub async fn execute_syscall_stage(&mut self) {
+        if !self.updates.hit_syscall && !matches!(self.blocked_on, BlockedOn::Syscall(_)) {
+            return;
+        }
+
+        // Determine if we should attempt to execute a new syscall or poll on a previous syscall
+        // the processor blocked on.
+        let syscall = match &self.blocked_on {
+            BlockedOn::Nothing => self.current_datapath.get_syscall_arguments(),
+            BlockedOn::Syscall(syscall) => syscall.clone(),
+        };
+
+        match syscall {
+            Syscall::Exit => {
+                self.current_datapath.halt();
+            }
+            Syscall::PrintInt(val) => {
+                self.scope
+                    .send(DatapathUpdate::System(SystemUpdate::Message(
+                        val.to_string(),
+                    )))
+                    .await
+                    .unwrap();
+            }
+            Syscall::PrintFloat(val) => {
+                self.scope
+                    .send(DatapathUpdate::System(SystemUpdate::Message(
+                        val.to_string(),
+                    )))
+                    .await
+                    .unwrap();
+            }
+            Syscall::PrintDouble(val) => {
+                self.scope
+                    .send(DatapathUpdate::System(SystemUpdate::Message(
+                        val.to_string(),
+                    )))
+                    .await
+                    .unwrap();
+            }
+            Syscall::PrintString(val) => {
+                self.scope
+                    .send(DatapathUpdate::System(SystemUpdate::Message(
+                        val.to_string(),
+                    )))
+                    .await
+                    .unwrap();
+            }
+            Syscall::ReadInt => {
+                let scan_result = self.scanner.next_int();
+                match scan_result {
+                    None => {
+                        self.blocked_on = BlockedOn::Syscall(syscall);
+                    }
+                    Some(scan_result) => {
+                        self.blocked_on = BlockedOn::Nothing;
+                        match self.current_datapath.as_datapath_ref() {
+                            DatapathRef::MIPS(_) => {
+                                self.current_datapath.set_register_by_str("v1", scan_result);
```

Comment:
I could be wrong about this but I think the register used for both storing the syscall code and receiving integer input is $v0

Reply:
I did a little research into this and I think I might move all of them somewhere else. The v registers are apparently supposed to be used for returned values and the a registers are supposed to be used for arguments. We can keep $f0 the argument/return register for the coprocessor just for convenience but I think we should switch the return location to $v0 and the syscall number to $a0, and its argument to $a1.